packages/lisk-core/src/channels/base.js

```diff
@@ -14,25 +11,30 @@ module.exports = class Base {
 	constructor(moduleAlias, events, actions, options = {}) {
 		this.moduleAlias = moduleAlias;
 		this.options = options;
+		this.eventsList = [];
+		this.actionsList = [];
 
-		eventsList.set(
-			this,
-			(options.skipInternalEvents ? events : internalEvents.concat(events)).map(
-				e => new Event(`${this.moduleAlias}:${e}`, null, null),
-			),
+		events.forEach((event) =>
+			this.eventList.push(new Event(`${moduleAlias}:${event}`, null, null)),
 		);
-		actionsList.set(
-			this,
-			actions.map(a => new Action(`${this.moduleAlias}:${a}`, null, null)),
+
+		if (!options.skipInternalEvents) {
+			internalEvents.forEach((event) =>
+				this.eventList.push(new Event(`${moduleAlias}:${event}`, null, null)),
+			);
+		}
+
+		actions.forEach((action) =>
+			this.actionsList.push(new Action(`${moduleAlias}:${action}`, null, null)),
```

Comment:
The problem with this change would be end user have direct access to `eventsList` by `channel.eventsList`. So user can modify the list which will cause problems while validating the events when publishing.

Once a module is loaded and channel instance is created, its list of events and actions should not be modified in any case, in any way. That was the intention of using WeakMap.

Reply:
Then in that case, we can use Object.freeze to make actionList and eventList arrays read-only.

Reply:
Yes, that would be possible. If we do not plan to extend actions or events after channel was created, we can make sure that it can not be redefined like below:

```js
const object1 = {
  property1: ["a", "b", "c"],
};

object1.property1 = Object.freeze(object1.property1);
Object.defineProperty(object1, "property1", { configurable: false, writable: false });

object1.property1 = ["0"];

console.log(object1.property1);
// expected output:  Array ["a", "b", "c"]
```

Reply:
So in coding practice, we still can use, no editor or intellisense will complaint on it. It will throw error on execution `Object property is not configurable ... ` , which is not user friendly, is it? 

```
object1.property1 = ["0"];
```

There are tons of articles suggested people are using this approach already in place, e.g. 

https://humanwhocodes.com/blog/2014/01/21/private-instance-members-with-weakmaps-in-javascript/
http://2ality.com/2016/01/private-data-classes.html
https://chrisrng.svbtle.com/using-weakmap-for-private-properties
https://www.sitepen.com/blog/legitimate-memory-efficient-privacy-with-es6-weakmaps/

I can't get your point stating in the issue `WeakMap usage is unnecessary`, while `WeakMap` is legitimate javascript feature and used to implement private data members. There is not other solution that can fully achieve this concept. 

Its clearly without current approach we can't get fully `private data members`, what benefits we will get in proposed approach of `Object.defineProperty`?

Reply:
I'm not against using WeekMaps, I'm against to have a global variable that holds the state of the instances. We can use WeakMap inside the class. I just didn't know that you wanted to make action and event lists immutable. 

Also the initial implementation also allows users to update the arrays since WeakMap.get returns the array reference and it's still possible to push new items.

```js
const myMap = new WeakMap();
const o1 = {};

myMap.set(o1, ["a","b"]);

const arr = myMap.get(o1);

arr.push("c");

console.log("updated", myMap.get(o1))
// "updated" Array ["a", "b", "c"]
```

Reply:
The intention is to have `private data member`, which ever approach you think is suitable and can be good for testing, go for it. Just keeping in mind it must be `private` and `immutable`.

Reply:
@nazarhussain If the intention is to have those variables as private as well, I agree that WeakMaps are a good way to go. I updated the code to have event & action lists private and immutable.